frontend-wasm/src/component/instance.rs

```diff
@@ -0,0 +1,138 @@
+use std::collections::HashMap;
+
+use miden_hir::cranelift_entity::PrimaryMap;
+
+use crate::{
+    component::Trampoline, error::WasmResult, module::module_env::ParsedModule, WasmError,
+};
+
+use super::{
+    ComponentTypes, CoreDef, GlobalInitializer, InstantiateModule, LinearComponent,
+    LinearComponentTranslation, LoweredIndex, RuntimeImportIndex, RuntimeInstanceIndex,
+    StaticModuleIndex, TypeFuncIndex,
+};
+
+/// A component import
+#[derive(Debug)]
+pub struct ComponentImport {
+    pub runtime_import_index: RuntimeImportIndex,
+    pub signature: TypeFuncIndex,
+}
+
+pub struct ComponentInstance<'data> {
+    pub modules: PrimaryMap<StaticModuleIndex, ParsedModule<'data>>,
+    pub module_instances: PrimaryMap<RuntimeInstanceIndex, StaticModuleIndex>,
+    pub component: LinearComponent,
+    pub component_types: ComponentTypes,
+    pub imports: HashMap<StaticModuleIndex, Vec<ComponentImport>>,
+}
+
+impl<'data> ComponentInstance<'data> {
+    pub fn ensure_module_names(&mut self) {
+        for (idx, parsed_module) in self.modules.iter_mut() {
+            parsed_module
+                .module
+                .set_name_fallback(format!("module{}", idx.as_u32()));
+        }
+    }
+
+    pub fn module(&self, idx: RuntimeInstanceIndex) -> &ParsedModule<'data> {
+        &self.modules[self.module_instances[idx]]
+    }
+}
+
+pub struct ComponentInstanceBuilder<'data> {
+    linear_component_translation: LinearComponentTranslation,
+    component_types: ComponentTypes,
+    modules: PrimaryMap<StaticModuleIndex, ParsedModule<'data>>,
+}
+
+impl<'data> ComponentInstanceBuilder<'data> {
+    pub fn new(
+        linear_component_translation: LinearComponentTranslation,
+        component_types: ComponentTypes,
+        modules: PrimaryMap<StaticModuleIndex, ParsedModule<'data>>,
+    ) -> Self {
+        Self {
+            linear_component_translation,
+            component_types,
+            modules,
+        }
+    }
+
+    pub fn build(self) -> WasmResult<ComponentInstance<'data>> {
+        let mut module_instances: PrimaryMap<RuntimeInstanceIndex, StaticModuleIndex> =
+            PrimaryMap::new();
+        let mut lower_imports: HashMap<LoweredIndex, RuntimeImportIndex> = HashMap::new();
+        let mut imports: HashMap<StaticModuleIndex, Vec<ComponentImport>> = HashMap::new();
+        let component = &self.linear_component_translation.component;
+        for initializer in &component.initializers {
+            match initializer {
+                GlobalInitializer::InstantiateModule(m) => {
+                    match m {
+                        InstantiateModule::Static(static_module_idx, args) => {
+                            if module_instances
+                                .values()
+                                .find(|idx| **idx == *static_module_idx)
+                                .is_some()
+                            {
+                                return Err(WasmError::Unsupported(format!(
```

Comment:
No reason. In terms of CM core module instance is a core module with fulfilled imports. I consider our IR `Module` as a core module instance counterpart since the imports are provided as MAST root hashes. I have not seen multiple core module instances "in the wild" so far in the basic wallet and p2id note script. I want to see one to understand the use case and the context. I thought until then, we could get away without the multiple core module instances support.

Good point with the documentation! I'll start with README and put a link to it in the mdbook.

Reply:
> I consider our IR Module as a core module instance counterpart since the imports are provided as MAST root hashes

That matches how I've been thinking about them too, so makes sense!

> I have not seen multiple core module instances "in the wild" so far in the basic wallet and p2id note script

After doing some more poking around the CM projects and discussions, it seems that in practice this isn't likely to come up for code compiled together as a unit, however it _can_ happen when tools produce components from a set of core Wasm modules, or that modify an existing component to "link" multiple core Wasm modules together. In any case, I think it's fine to not worry about supporting this until we have a use case where it comes up, but we should probably treat it as a future extension of the frontend, so handle it as a soft error with diagnostics and documentation and all of that now (which aside from docs is already done here), and maybe an issue to track it as a future roadmap item.

Reply:
I agree.

Reply:
I made a README and listed the unsupported features and the limitations of the frontend with explanations where necessary and linked it to the book in https://github.com/0xPolygonMiden/compiler/pull/100/commits/4b3d4e52a4ce955e4b90fbe623d244f670de497f